src/programming-assistants.rst

```diff
@@ -1,23 +1,179 @@
+
+-- Insert Diagram Here --
+
+A synthesizer produces a 'candidate program' based on the specifications provided and sends the program to a verifier to validate that the program fully satisfies the specifications. If the verifier successfully validates the program, no further action is taken. However, if the verifier determines that the candidate is not satisfactory, it provides feedback to the synthesizer to be used to guide the development of a new candidate program. This process is repeated until the candidate program passes all of the specification.
+
+Gvero and Kuncak have a proposed a new code assistance tool, anyCode, for integrated development environments (IDEs) that produces a list of syntheses based on user-entered free-form queries :cite:`gvero2015synthesizing`. The queries contain a mixture of English and Java syntax and produce expressions that account for the correct syntax, types and scoping rules of Java.
+
+Given the query `open file example.txt`, anyCode would return a list of potential Java expressions:
+.. code-block:: Java
+  File file = new File("/Users/example/example.txt");
+  BufferedReader reader = new BufferedReader(new FileReader("/Users/example/example.txt"));
+  
+The authors state that the key feature that differentiates their tool is that it does not take into account any prior code context. The main take aways of their tool are that it often produces the expected code fragments, tolerates the flexibility of the English language, and it can repair incorrect Java expressions that use the wrong syntax or missing arguments.
+
+anyCode was evaluated with 45 examples where the effectiveness criteria was that it was able to successfully synthesize expected expressions and to list them among the top 10 solutions. The two models used to build the tool, unigram and PCFG, were evaluated separately and together. Using only the unigram model, the expected expression was only in the top 10 27% of the time and 18% using the PCFG model.
+
+However, when both models were used, the expected expression was in the top 10 82% of time and was the #1 option 44% of the time. More implementation details of the two models can be found `here <https://lara.epfl.ch/~kuncak/papers/GveroKuncak15SynthesizingJavaExpressionsFreeFormQueries.pdf>`_. anyCode was given an average runtime of 60 ms to generate the top 10 expressions, proving that it can effectively and efficiently synthesize expressions.
+
 Code Completion
 ---------------
 
+**Code Completion**, a subsection of Program Synthesis, refers to the process of intelligent completion suggestion making for lines of code a user writes during the development of an application. It is intended to speed up the process of coding applications by reducing typos and other mistakes. Code Completion assistants serves as convenient way to access function syntax and descriptions, significantly reducing the amount of memorization required. 
```

Comment:
This isn't the clearest definition of code completion. I'd say something about how code completion systems are simple program synthesis systems that generate code snippets for insertion at the cursor (whereas other program synthesis systems might fill multiple holes).

Reply:
By that definition, though, the Raychev 2014 system isn't code completion.

It might be better to just define "code completion system" as a phrase that means "program synthesis integrated into the editor".